src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                                }
+
+                                byte first = _buffer[_consumed];
+
+                                if (first <= JsonConstants.Space)
+                                {
+                                    SkipWhiteSpace();
+                                    if (!HasMoreData())
+                                        return false;
+                                    first = _buffer[_consumed];
+                                }
+
+                                return ConsumeValue(first);
+                            }
+                            return false;
+                        }
+                    }
+                }
+
+                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.ExpectedStartOfValueNotFound, marker);
+            }
+            return true;
+        }
+
+        // Consumes 'null', or 'true', or 'false'
+        private bool ConsumeLiteral(ReadOnlySpan<byte> literal, JsonTokenType tokenType)
+        {
+            ReadOnlySpan<byte> span = _buffer.Slice(_consumed);
+            Debug.Assert(span.Length > 0);
+            Debug.Assert(span[0] == 'n' || span[0] == 't' || span[0] == 'f');
+
+            if (!span.StartsWith(literal))
+                return CheckLiteral(span, literal);
+
+            ValueSpan = span.Slice(0, literal.Length);
+            _tokenType = tokenType;
+            _consumed += literal.Length;
+            _lineBytePosition += literal.Length;
+            return true;
+        }
+
+        private bool CheckLiteral(ReadOnlySpan<byte> span, ReadOnlySpan<byte> literal)
+        {
+            Debug.Assert(span.Length > 0 && span[0] == literal[0]);
+
+            int indexOfFirstMismatch = 0;
+
+            for (int i = 1; i < literal.Length; i++)
+            {
+                if (span.Length > i)
+                {
+                    if (span[i] != literal[i])
+                    {
+                        _lineBytePosition += i;
+                        ThrowInvalidLiteral(span);
+                    }
+                }
+                else
+                {
+                    indexOfFirstMismatch = i;
+                    break;
+                }
+            }
+
+            Debug.Assert(indexOfFirstMismatch > 0 && indexOfFirstMismatch < literal.Length);
+
+            if (IsLastSpan)
+            {
+                _lineBytePosition += indexOfFirstMismatch;
+                ThrowInvalidLiteral(span);
+            }
+            return false;
+        }
+
+        private void ThrowInvalidLiteral(ReadOnlySpan<byte> span)
+        {
+            byte firstByte = span[0];
+            ExceptionResource resource = ExceptionResource.ExpectedNull;
+            if (firstByte == 't')
+                resource = ExceptionResource.ExpectedTrue;
+            else if (firstByte == 'f')
+                resource = ExceptionResource.ExpectedFalse;
+            ThrowHelper.ThrowJsonReaderException(ref this, resource, bytes: span);
```

Comment:
Nit: consider:
```C#
ExceptionResource resource;
switch (span[0])
{
    case 't': resource = ExceptionResource.ExpectedTrue; break;
    case 'f': resource = ExceptionResource.ExpectedFalse; break;
    default: resource = ExceptionResource.ExpectedNull; break;
}
ThrowHelper.ThrowJsonReaderException(ref this, resource, bytes: span); 
```
I personally find that easier to read.
}

Reply:
I'd do

```C#
switch (span[0])
{
    case 't': resource = ExceptionResource.ExpectedTrue; break;
    case 'f': resource = ExceptionResource.ExpectedFalse; break;
    default:
        Debug.Assert(span[0] == 'n');
        resource = ExceptionResource.ExpectedNull;
        break;
}
ThrowHelper.ThrowJsonReaderException(ref this, resource, bytes: span); 
```

But that's because I don't like accidental defaults.

I'd probably also not elide the newlines.  Because that's me.

Reply:
(Adding the "assert it's one of these three things" at the beginning of the method would then make Steve's condensed thing better)